File: 329. Longest Increasing Path in a Matrix.cpp
class Solution {
public:
    int trav(vector<vector<int>>& matrix, vector<vector<int>>& memo, int lastEle, int i, int j, int n, int m) {
        if (i < 0 || j < 0 || i >= n || j >= m) return 0;

        if (lastEle >= matrix[i][j]) return 0;
        if (memo[i][j] != -1) return memo[i][j];

        int found = 1 + max({
            trav(matrix, memo, matrix[i][j], i + 1, j, n, m),
            trav(matrix, memo, matrix[i][j], i - 1, j, n, m),
            trav(matrix, memo, matrix[i][j], i, j + 1, n, m),
            trav(matrix, memo, matrix[i][j], i, j - 1, n, m)
        });

        memo[i][j] = found;
        return found;
    }

    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        int maxPath = 1;
        vector<vector<int>> memo(n, vector<int>(m, -1));

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (memo[i][j] != -1) {
                    maxPath = max(maxPath, memo[i][j]);
                    continue;
                }

                int found = 1 + max({
                    trav(matrix, memo, matrix[i][j], i + 1, j, n, m),
                    trav(matrix, memo, matrix[i][j], i - 1, j, n, m),
                    trav(matrix, memo, matrix[i][j], i, j + 1, n, m),
                    trav(matrix, memo, matrix[i][j], i, j - 1, n, m)
                });

                memo[i][j] = found;
                maxPath = max(maxPath, found);
            }
        }

        return maxPath;
    }
};

File: 329. Longest Increasing Path in a Matrix.java
class Solution {
    public int dfs(int[][] matrix, int i, int j, int past, int[][] dp, boolean[][] visited){
        if(i >= matrix.length || j >= matrix[0].length || i < 0 || j < 0 || matrix[i][j] <= past || visited[i][j]) return 0;

        if(dp[i][j] != -1) return dp[i][j];
        visited[i][j] = true;

        int res = 0;
        res = Math.max(res, 1 + dfs(matrix, i + 1, j, matrix[i][j], dp, visited));
        res = Math.max(res, 1 + dfs(matrix, i - 1, j, matrix[i][j], dp, visited));
        res = Math.max(res, 1 + dfs(matrix, i, j + 1, matrix[i][j], dp, visited));
        res = Math.max(res, 1 + dfs(matrix, i, j - 1, matrix[i][j], dp, visited));

        visited[i][j] = false;

        return dp[i][j] = res;
    }
    public int longestIncreasingPath(int[][] matrix) {
        int n = matrix.length, m = matrix[0].length;
        int[][] dp = new int[n][m];
        boolean[][] visited = new boolean[n][m];

        for(int i = 0; i < n; i++) Arrays.fill(dp[i], -1);

        int ans = 1;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                ans = Math.max(ans, dfs(matrix, i, j, -1, dp, visited));
            }
        }

        return ans;
    }
}